worker/caasadmission/admission.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasadmission
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"github.com/juju/errors"
+	admission "k8s.io/api/admissionregistration/v1beta1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/juju/juju/caas/kubernetes/provider"
+)
+
+type CertificateBroker interface {
+	// RegisterDomainLeases registers a new domain name lease on the broker's
+	// certificate. Returns a cleanup lease function on success.
+	RegisterDomainLeases(...string) (func(), error)
+
+	// RootCertificates set of root certificates in pem format for the broker.
+	RootCertificates() ([]byte, error)
+}
+
+// TODO this is a tmp struct and won't make it to final
+type CertWatcherBroker struct {
+	CertWatcher func() *tls.Certificate
+}
+
+// Represents a creator of mutating webhooks that is context aware of the
+// current controller.
+type AdmissionCreator interface {
+	EnsureMutatingWebhookConfiguration() (func(), error)
+}
+
+// Func type of AdmissionCreator
+type AdmissionCreatorFunc func() (func(), error)
+
+const (
+	Component = "admission"
+)
+
+var (
+	anyMatch = []string{"*"}
+)
+
+// EnsureMutatingWebhookConfiguration implements AdmissionCreator interface for
+// func type
+func (a AdmissionCreatorFunc) EnsureMutatingWebhookConfiguration() (func(), error) {
+	return a()
+}
+
+// NewAdmissionCreator instantiates a new AdmissionCreator for the supplied
+// context arguments.
+func NewAdmissionCreator(
+	certBroker CertificateBroker,
+	namespace, modelName string,
```

Comment:
same here, do we need both?

Reply:
yes I want to be explicit about namespace and model name are not tied together otherwise we are repeating logic and that could undo us at some stage.